File: LeetCode/Python Solutions/Check if There is a Valid Partition For The Array/check.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        total_nums = len(nums)
        
        @cache
        def canPartitionFrom(index: int) -> bool:
            # Base case: If we've reached the end of the list, return True
            if index == total_nums:
                return True

            # Check for a pair of equal numbers
            if index + 1 < total_nums and nums[index] == nums[index + 1] and canPartitionFrom(index + 2):
                return True

            # Check for a triplet of equal numbers
            if index + 2 < total_nums and nums[index] == nums[index + 1] == nums[index + 2] and canPartitionFrom(index + 3):
                return True

            # Check for a triplet of consecutive numbers
            if index + 2 < total_nums and nums[index] + 1 == nums[index + 1] and nums[index + 1] + 1 == nums[index + 2] and canPartitionFrom(index + 3):
                return True

            return False

        return canPartitionFrom(0)

File: LeetCode/Java Solutions/Check if There is a Valid Partition For The Array/check.java
import java.util.HashMap;

public class Solution {
    private int[] nums;
    private HashMap<Integer, Boolean> memo;

    public boolean validPartition(int[] nums) {
        this.nums = nums;
        this.memo = new HashMap<>();
        return canPartitionFrom(0);
    }

    private boolean canPartitionFrom(int index) {
        if (index == nums.length) {
            return true;
        }

        if (memo.containsKey(index)) {
            return memo.get(index);
        }

        // Check for a pair of equal numbers
        if (index + 1 < nums.length && nums[index] == nums[index + 1] && canPartitionFrom(index + 2)) {
            memo.put(index, true);
            return true;
        }

        // Check for a triplet of equal numbers
        if (index + 2 < nums.length && nums[index] == nums[index + 1] && nums[index] == nums[index + 2] && canPartitionFrom(index + 3)) {
            memo.put(index, true);
            return true;
        }

        // Check for a triplet of consecutive numbers
        if (index + 2 < nums.length && nums[index] + 1 == nums[index + 1] && nums[index + 1] + 1 == nums[index + 2] && canPartitionFrom(index + 3)) {
            memo.put(index, true);
            return true;
        }

        memo.put(index, false);
        return false;
    }
}

